lib/Dialect/FIRRTL/Transforms/InferDomains.cpp

```diff
@@ -0,0 +1,1426 @@
+}
+
+/// From a domain info attribute, get the row of associated domains for a
+/// hardware value at index i.
+static auto getPortDomainAssociation(ArrayAttr info, size_t i) {
+  if (info.empty())
+    return info.getAsRange<IntegerAttr>();
+  return cast<ArrayAttr>(info[i]).getAsRange<IntegerAttr>();
+}
+
+/// Return true if the value is a port on the module.
+static bool isPort(FModuleOp module, BlockArgument arg) {
+  return arg.getOwner()->getParentOp() == module;
+}
+
+/// Return true if the value is a port on the module.
+static bool isPort(FModuleOp module, Value value) {
+  auto arg = dyn_cast<BlockArgument>(value);
+  if (!arg)
+    return false;
+  return isPort(module, arg);
+}
+
+//====--------------------------------------------------------------------------
+// Circuit-wide state.
+//====--------------------------------------------------------------------------
+
+/// Each declared domain in the circuit is assigned an index, based on the order
+/// in which it appears. Domain associations for hardware values are represented
+/// as a list of domains, sorted by the index of the domain type.
+using DomainTypeID = size_t;
+
+/// Information about the domains in the circuit. Able to map domains to their
+/// type ID, which in this pass is the canonical way to reference the type
+/// of a domain.
+namespace {
+struct CircuitDomainInfo {
+  CircuitDomainInfo(CircuitOp circuit) { processCircuit(circuit); }
+
+  ArrayRef<DomainOp> getDomains() const { return domainTable; }
+  size_t getNumDomains() const { return domainTable.size(); }
+  DomainOp getDomain(DomainTypeID id) const { return domainTable[id]; }
+
+  DomainTypeID getDomainTypeID(DomainOp op) const {
+    return typeIDTable.at(op.getNameAttr());
+  }
+
+  DomainTypeID getDomainTypeID(StringAttr name) const {
+    return typeIDTable.at(name);
+  }
+
+  DomainTypeID getDomainTypeID(FlatSymbolRefAttr ref) const {
+    return getDomainTypeID(ref.getAttr());
+  }
+
+  DomainTypeID getDomainTypeID(ArrayAttr info, size_t i) const {
+    auto name = getDomainPortTypeName(info, i);
+    return getDomainTypeID(name);
+  }
+
+  DomainTypeID getDomainTypeID(Value value) const {
+    assert(isa<DomainType>(value.getType()));
+    if (auto arg = dyn_cast<BlockArgument>(value)) {
+      auto *block = arg.getOwner();
+      auto *owner = block->getParentOp();
+      auto module = cast<FModuleOp>(owner);
+      auto info = module.getDomainInfoAttr();
+      auto i = arg.getArgNumber();
+      return getDomainTypeID(info, i);
+    }
+
+    auto result = dyn_cast<OpResult>(value);
+    auto *owner = result.getOwner();
+    auto instance = cast<InstanceOp>(owner);
+    auto info = instance.getDomainInfoAttr();
+    auto i = result.getResultNumber();
+    return getDomainTypeID(info, i);
+  }
+
+private:
+  void processDomain(DomainOp op) {
+    auto index = domainTable.size();
+    auto name = op.getNameAttr();
+    domainTable.push_back(op);
+    typeIDTable.insert({name, index});
```

Comment:
Mega-nit: consider changing the order of modifications to `domainTable` and `typeIDTable` to move the temporaries closer to their single use sites.

Also, as always, consider inlining single use variables when it helps clarity (or using comments: `{*/name=*/op.getNameAttr(), /*index=*/domainTable.size()}`.